indico/modules/legal/controllers.py

```diff
@@ -5,22 +5,38 @@
 # modify it under the terms of the MIT License; see the
 # LICENSE file for more details.
 
-from flask import redirect
+from flask import flash, redirect, request, session
 
 from indico.modules.admin import RHAdminBase
 from indico.modules.legal import legal_settings
-from indico.modules.legal.forms import LegalMessagesForm
-from indico.modules.legal.views import WPDisplayPrivacyPolicy, WPDisplayTOS, WPManageLegalMessages
+from indico.modules.legal.forms import AgreementForm, LegalMessagesForm
+from indico.modules.legal.views import WPDisplayAgreement, WPDisplayPrivacyPolicy, WPDisplayTOS, WPManageLegalMessages
+from indico.util.date_time import now_utc
+from indico.util.i18n import _
 from indico.web.flask.util import url_for
 from indico.web.rh import RH
+from indico.web.util import url_for_index
 
 
 class RHManageLegalMessages(RHAdminBase):
     def _process(self):
-        form = LegalMessagesForm(**legal_settings.get_all())
+        settings = legal_settings.get_all()
+        form = LegalMessagesForm(**settings)
+
         if form.validate_on_submit():
-            legal_settings.set_multi(form.data)
+            data = form.data
+
+            has_privacy_policy = data['privacy_policy'] or data['privacy_policy_url']
+            has_tos = data['tos'] or data['tos_url']
+
+            if data['terms_require_accept'] and not has_privacy_policy and not has_tos:
+                flash(_('Requiring to accept terms requires a privacy policy '
+                        'and/or terms of service being set'), 'error')
+                return redirect(url_for('legal.manage'))
```

Comment:
I would move this into the `LegalMessagesForm` and fail nicely on the validation level. Flash+redirect means any data changed/entered in the form is lost